### PR TITLE
feat(wallet) Implement Meld Buy Flow in the FE

### PR DESCRIPTION
**Pull request by muliswilliam**
## Description 

Introduces a new `Buy Crypto` flow using `Meld` API's in the `Brave Wallet`.
User are now able to get realtime `Quotes` from multiple providers with more assets to choose from.

Resolves <https://github.com/brave/brave-browser/issues/38985>

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
1. Open the `Wallet` and navigate to the `Buy` screen
2. `Ethereum` should be automatically pre-selected and should fetch `quotes`
3. Test selecting other `tokens`, changing `account` and changing `currencies`
4. Test filtering by different `Payment Methods`
5. Go back to the `Portfolio` page and click the `More Menu` for a token, if `Buy` is available for that token you should see a `Buy` button and it should pre populate the `Token` and proper `Account` after clicking.

https://github.com/user-attachments/assets/877cfb89-79ae-4ce1-87ae-18c589799ece

### CODE REVIEW

**Review by yrliou — APPROVED** [review]
core++

File: components/resources/wallet_strings.grdp [hunk 1]
```diff
@@ -618,6 +618,23 @@
   <message name="IDS_BRAVE_WALLET_SELL_WITH_PROVIDER" desc="Sell with Provider option label">Sell with <ph name="OFFRAMP_PROVIDER">$1<ex>Ramp</ex></ph></message>
   <message name="IDS_BRAVE_WALLET_BUY_WITH_SARDINE" desc="Buy with Sardine option label">Buy with Sardine</message>
   <message name="IDS_BRAVE_WALLET_BUY_DISCLAIMER" desc="Brave disclaimer text in buy screen">Financial and transaction data is processed by our onramp partners. Brave does not collect or have access to such data.</message>
+  <message name="IDS_BRAVE_WALLET_TRANSACTIONS_PARTNER" desc="Transactions partner modal title">Transactions partner</message>
+  <message name="IDS_BRAVE_WALLET_TRANSACTION_PARTNER_CONSENT" desc="Meld consent description">Buying or selling crypto in Brave Wallet uses Meld.io — an on-ramp/off-ramp aggregator that provides a smooth experience and the best available pricing. Your information will be shared with Meld.io to complete the transaction.</message>
```
Comment (mkarolin) [inline components/resources/wallet_strings.grdp:622]: Should we be making the claim about the best available pricing? These types of things usually come with a disclaimer.

Comment (Douglashdaniel) [reply components/resources/wallet_strings.grdp:622]: cc: @jamesmudgett @kleantzogu

Comment (Douglashdaniel) [reply components/resources/wallet_strings.grdp]: @mkarolin we have updated this copy here https://github.com/brave/brave-core/compare/3f29b5d103ed154938a55caad7d0dd7726744356..def15c35ca71d06f687e85880eff6d9a11f7153b

**Review by mkarolin — APPROVED** [review]
`strings`++

File: components/brave_wallet_ui/utils/string-utils.ts [hunk 2]
```diff
@@ -190,12 +190,12 @@ export const capitalizeFirstLetter = (input: string) => {
...
+    }
 
-  const firstHalf = integerString.slice(0, 3)
-  const secondHalf = integerString.slice(-3)
-  const reduced = firstHalf.concat('...', secondHalf)
-  return reduced
+    const firstHalf = integerString.slice(0, 3)
+    const secondHalf = integerString.slice(-3)
+    const reduced = firstHalf.concat('...', secondHalf)
+    return reduced
```
[the remark this answers is not in the document]
Comment (Douglashdaniel) [reply components/brave_wallet_ui/utils/string-utils.ts]: This is fixed!

**Comment by kdenhartog** [comment]
just noting I see consent text in the PR, but don't see any in the video provided. I'm going to build this and take a look at it tomorrow to see if we're good to go for privacy/security or if there's some changes we might still need.

**Comment by Douglashdaniel** [comment]
> just noting I see consent text in the PR, but don't see any in the video provided. I'm going to build this and take a look at it tomorrow to see if we're good to go for privacy/security or if there's some changes we might still need.

@kylehickinson indeed, I had previously accepted the consent form so it was not appearing in the video. This is how it will be displayed. note(I'll update this PR to link to https://www.meld.io/terms-of-use as it looks like it was currently linking to https://brave.com/terms-of-use/.

https://github.com/user-attachments/assets/13b7b5db-6f10-4b93-a161-1bcbe819180a

**Comment by kdenhartog** [comment]
- [ ] The terms of service link seems to be not working properly in the panel. The one in `brave://wallet/crypto/portfolio/assets` is working as expected though.

meld terms of service link


- [ ] nit: non security blocking. Do we need to set a default here? It wasn't clear to me when I first was looking at displayed prices that I was on US and that each country would have different values. This may artificially limit the buying options to the default country which may end up not presenting the best prices for the user.
Screenshot 2024-09-13 at 4 49 47 PM

- [ ] it looks like `Local Cash Vouchers` and `Local Manual Bank Transfer` are listed quite a bit here.
Screenshot 2024-09-13 at 4 51 54 PM. 

- [ ] Also do we need to list all payment methods if they're not supported by a particular country? For example, `UK Faster Payments` seems UK specific so do we need to list it if I've filtered to New Zealand? In many cases, most of these aren't relevant to the user because the payment processor isn't supported in that particular country. Additionally, it would be useful to have an option to select "all" rather than a specific method for this.

 - [ ] nit: there's a lot of failed price fetching happening. Are we supposed to be requesting this many prices all at once or is there a more optimal way that we could be doing this?
console errors. Also, to what degree are we caching response for this?

- [ ] The Buy Button isn't displaying as expected when clicked.
buy buttons unexpected

- [ ] Related to above, when clicking some of the payment processors (Sardine for example) nothing shows with shields up because they've got analytics that are being blocked and crashing the page. Meld will likely need to fix this

- [ ] What appears to be a static JWT token is being set as a query parameter when being redirected to meldcrypto.com. Can you provide additional details about what this token is used for? 

- [ ] The default asset should be reset depending on which account is selected. We shouldn't be passing in a filecoin address when we want to be purchasing ETH as an example.
wrong account passed in

File: components/brave_wallet_ui/stories/locale.ts [hunk 3]
```diff
@@ -668,6 +668,30 @@ provideStrings({
   braveWalletBuyDisclaimer:
     'Financial and transaction data is processed by our onramp partners. ' +
     'Brave does not collect or have access to such data.',
+  braveWalletTransactionsPartner: 'Transactions partner',
+  braveWalletTransactionPartnerConsent:
+    'Brave Wallet uses Meld.io to help aggregate and surface various ' +
+    'crypto providers for your region. Some of your information will be ' +
+    'shared with Meld.io to complete the transaction. For more information ' +
+    'please read Meld’s terms of use.',
```
Comment (kdenhartog) [inline components/brave_wallet_ui/stories/locale.ts:676]:
```suggestion
    'Brave Wallet uses Meld.io to help aggregate and surface various ' +
    'crypto providers for your region. Some of your information including ' +
    'your wallet address and country code will be shared with Meld.io to ' + 
    'complete the transaction. For more information please read Meld’s ' + 
    'terms of use.',
```

File: components/resources/wallet_strings.grdp [hunk 4]
```diff
@@ -618,6 +618,23 @@
   <message name="IDS_BRAVE_WALLET_SELL_WITH_PROVIDER" desc="Sell with Provider option label">Sell with <ph name="OFFRAMP_PROVIDER">$1<ex>Ramp</ex></ph></message>
   <message name="IDS_BRAVE_WALLET_BUY_WITH_SARDINE" desc="Buy with Sardine option label">Buy with Sardine</message>
   <message name="IDS_BRAVE_WALLET_BUY_DISCLAIMER" desc="Brave disclaimer text in buy screen">Financial and transaction data is processed by our onramp partners. Brave does not collect or have access to such data.</message>
+  <message name="IDS_BRAVE_WALLET_TRANSACTIONS_PARTNER" desc="Transactions partner modal title">Transactions partner</message>
+  <message name="IDS_BRAVE_WALLET_TRANSACTION_PARTNER_CONSENT" desc="Meld consent description">Brave Wallet uses Meld.io to help aggregate and surface various crypto providers for your region. Some of your information will be shared with Meld.io to complete the transaction. For more information please read Meld's terms of use.</message>
```
Comment (kdenhartog) [inline components/resources/wallet_strings.grdp:622]:
```suggestion
  <message name="IDS_BRAVE_WALLET_TRANSACTION_PARTNER_CONSENT" desc="Meld consent description>Brave Wallet uses Meld.io to help aggregate and surface various crypto providers for your region. Some of your information including your wallet address and country code will be shared with Meld.io to complete the transaction. For more information please read Meld’s terms of use.</message>
```

**Review by kdenhartog — COMMENTED** [review]
no code specific concerns from what I'm seeing. I think this is fine from a security perspective and made a suggestion for consent notice that I think should be fine.

File: components/brave_wallet_ui/stories/locale.ts (hunk 3 above)
Comment (Douglashdaniel) [reply components/brave_wallet_ui/stories/locale.ts]: Updated with https://github.com/brave/brave-core/compare/1dc018b8bcdd9b4a4caab31961b27e377817ae32..8c5fddaa6ddfbdb5cf22c0500b0917ac87851bfb

Screenshot 1

File: components/resources/wallet_strings.grdp (hunk 4 above)
Comment (Douglashdaniel) [reply components/resources/wallet_strings.grdp]: Updated with https://github.com/brave/brave-core/compare/1dc018b8bcdd9b4a4caab31961b27e377817ae32..8c5fddaa6ddfbdb5cf22c0500b0917ac87851bfb

Screenshot 1

**Comment by Douglashdaniel** [comment]
@kdenhartog I have address most of your comments. The other issues not mentioned in this comment will need to be address by `Meld` themselves I believe.

> The terms of service link seems to be not working properly in the panel. The one in brave://wallet/crypto/portfolio/assets is working as expected though.

This should be fixed now with https://github.com/brave/brave-core/compare/1dc018b8bcdd9b4a4caab31961b27e377817ae32..8c5fddaa6ddfbdb5cf22c0500b0917ac87851bfb

https://github.com/user-attachments/assets/a3806e09-72bd-4f68-ac01-de7217bc3c6c

---

> The default asset should be reset depending on which account is selected. We shouldn't be passing in a filecoin address when we want to be purchasing ETH as an example.

This should be fixed now with https://github.com/brave/brave-core/compare/1dc018b8bcdd9b4a4caab31961b27e377817ae32..8c5fddaa6ddfbdb5cf22c0500b0917ac87851bfb
It will now filter `accounts` by `coinType` as to never allow selecting an unsupported `account`.

https://github.com/user-attachments/assets/35e49559-eb46-4669-b836-0b6141d191db

---

> nit: non security blocking. Do we need to set a default here? It wasn't clear to me when I first was looking at displayed prices that I was on US and that each country would have different values. This may artificially limit the buying options to the default country which may end up not presenting the best prices for the user.

> it looks like Local Cash Vouchers and Local Manual Bank Transfer are listed quite a bit here.

> Also do we need to list all payment methods if they're not supported by a particular country? For example, UK Faster Payments seems UK specific so do we need to list it if I've filtered to New Zealand? In many cases, most of these aren't relevant to the user because the payment processor isn't supported in that particular country.

These 3 should all be fixed now with https://github.com/brave/brave-core/compare/1dc018b8bcdd9b4a4caab31961b27e377817ae32..8c5fddaa6ddfbdb5cf22c0500b0917ac87851bfb `Payment Methods` are now filtered by `CountryCode` and `CurrencyCode`.

image | image
--|--

---

> Additionally, it would be useful to have an option to select "all" rather than a specific method for this.

Initially we were going to have a `checkbox` filter to select multiple `PaymentMethods` but `Meld` API only allows to pass `1`.

---

> nit: there's a lot of failed price fetching happening. Are we supposed to be requesting this many prices all at once or is there a more optimal way that we could be doing this? Also, to what degree are we caching response for this?

We are currently fetching all these token `Prices` in batches and values are being `cached`. Small issue currently out of our control is that the `Meld` token list does not include a `coingeckoId` for each token. So we are having to query prices using the tokens `symbol` which is not always reliable thus the amount of failed prices.

**Comment by kdenhartog** [comment]
>  What appears to be a static JWT token is being set as a query parameter when being redirected to meldcrypto.com. Can you provide additional details about what this token is used for?

Answering my own question here - this is a session token used to connect with meld. It's not a concern if this is taken and also is controlled by meld in how they choose to receive it.

> Related to above, when clicking some of the payment processors (Sardine for example) nothing shows with shields up because they've got analytics that are being blocked and crashing the page. Meld will likely need to fix this

I've opted to work with the Meld team to help address this, so nothing further blocking for you @Douglashdaniel. Thanks for the fixes

**Review by kdenhartog — APPROVED** [review]

**Review by onyb — APPROVED** [review]
++

**Pull request merged**